hcipy/optics/segmented_mirror.py

```diff
@@ -0,0 +1,160 @@
+import os
+import numpy as np
+import matplotlib.pyplot as plt
+
+from .optical_element import OpticalElement
+from ..field import Field
+from ..plotting import imshow_field
+
+
+class SegmentedMirror(OpticalElement):
+	"""A segmented mirror from a segmented aperture.
+
+	Parameters:
+	----------
+	indexed_aperture : Field
+		The *indexed* segmented aperture of the mirror, all pixels each segment being filled with its number for
+		segment identification. Segment gaps must be strictly zero.
+	seg_pos : CartesianGrid(UnstructuredCoords)
+		Segment positions of the aperture.
+	"""
+
+	def __init__(self, indexed_aperture, seg_pos):
+		self.ind_aper = indexed_aperture
+		self.segnum = len(seg_pos.x)
+		self.segmentlist = np.arange(1, self.segnum + 1)
+		self._coef = np.zeros((self.segnum, 3))
+		self.seg_pos = seg_pos
+		self.input_grid = indexed_aperture.grid
+		self._last_npix = np.nan  # see _setup_grids for this
+		self._surface = None
+
+	def forward(self, wavefront):
+		"""Propagate a wavefront through the segmented mirror.
+
+		Parameters
+		----------
+		wavefront : Wavefront
+			The incoming wavefront.
+
+		Returns
+		-------
+		Wavefront
+			The reflected wavefront.
+		"""
+		wf = wavefront.copy()
+		wf.electric_field *= np.exp(2j * self.surface * wavefront.wavenumber)
+		return wf
+
+	def backward(self, wavefront):
+		"""Propagate a wavefront backwards through the deformable mirror.
+
+		Parameters
+		----------
+		wavefront : Wavefront
+			The incoming wavefront.
+
+		Returns
+		-------
+		Wavefront
+			The reflected wavefront.
+		"""
+		wf = wavefront.copy()
+		wf.electric_field *= np.exp(-2j * self.surface * wavefront.wavenumber)
+		return wf
+
+	@property
+	def surface(self):
+		""" The surface of the segmented mirror in meters, the full surface as a Field.
+		"""
+		if self._surface is None:
+			self._surface = self.apply_coef()
+		return self._surface
+
+	@property
+	def coef(self):
+		""" The surface shape of the deformable mirror, in meters and radians; PTT segment coefficients.
+		"""
+		self._surface = None
+		return self._coef
+
+	def show_numbers(self):
+		""" Display the mirror pupil with numbered segments.
+		"""
+		imshow_field(self.ind_aper)
+		for i, par in enumerate(self.seg_pos):
+			plt.annotate(s=i+1, xy=par, xytext=par, color='white', fontweight='bold') #TODO: scale text size by segment size
```

Comment:
The text positions themselves need to be scaled by the aperture grid diameter.

This display function will work all right if the segment positions are in the same units like the input indexed aperture. I ran into issues because I would evaluate the indexed aperture on a differently sized grid than it was created with, while the segments positions are tied to the original grid it was created with. Not sure how to limit this properly - also, it is just supposed to be a convenience function.